File: backend/src/main/java/dev/taskflow/domain/model/TaskStatus.java
package dev.taskflow.domain.model;

import dev.taskflow.domain.exception.InvalidTaskTransitionException;

public enum TaskStatus {

    BACKLOG,
    TODO,
    IN_PROGRESS,
    IN_REVIEW,
    DONE,
    CANCELLED;

    public boolean canTransitionTo(TaskStatus target) {
        return switch (this) {
            case BACKLOG     -> target == TODO || target == CANCELLED;
            case TODO        -> target == IN_PROGRESS || target == BACKLOG || target == CANCELLED;
            case IN_PROGRESS -> target == IN_REVIEW || target == TODO || target == CANCELLED;
            case IN_REVIEW   -> target == DONE || target == IN_PROGRESS || target == CANCELLED;
            case DONE, CANCELLED -> false;
        };
    }

    public void validateTransitionTo(TaskStatus target) {
        if (!canTransitionTo(target)) {
            throw new InvalidTaskTransitionException(this, target);
        }
    }

    public boolean isTerminal() {
        return this == DONE || this == CANCELLED;
    }
}

File: frontend/src/app/shared/models/task-status.ts
export enum TaskStatus {
  BACKLOG = 'BACKLOG',
  TODO = 'TODO',
  IN_PROGRESS = 'IN_PROGRESS',
  IN_REVIEW = 'IN_REVIEW',
  DONE = 'DONE',
  CANCELLED = 'CANCELLED',
}

export const TASK_STATUS_TRANSITIONS: Record<TaskStatus, TaskStatus[]> = {
  [TaskStatus.BACKLOG]: [TaskStatus.TODO, TaskStatus.CANCELLED],
  [TaskStatus.TODO]: [TaskStatus.IN_PROGRESS, TaskStatus.BACKLOG, TaskStatus.CANCELLED],
  [TaskStatus.IN_PROGRESS]: [TaskStatus.IN_REVIEW, TaskStatus.TODO, TaskStatus.CANCELLED],
  [TaskStatus.IN_REVIEW]: [TaskStatus.DONE, TaskStatus.IN_PROGRESS, TaskStatus.CANCELLED],
  [TaskStatus.DONE]: [],
  [TaskStatus.CANCELLED]: [],
};

export function canTransition(from: TaskStatus, to: TaskStatus): boolean {
  return TASK_STATUS_TRANSITIONS[from].includes(to);
}
